Fix: 4xx and 5xx responses from API Gateway cause Invicti findings (DataBiosphere/azul-private#154)

dsotirho-ucsc opened the pull request:
<!--
This is the PR template for regular PRs against `develop`. Edit the URL in your
browser's location bar, appending either `&template=anvilprod-promotion.md`,
`&template=prod-promotion.md`, `&template=anvilprod-hotfix.md`, `&template=prod-
hotfix.md`, `&template=backport.md` or `&template=upgrade.md` to switch the
template.
-->

Connected issues: DataBiosphere/azul-private#154


## Checklist


### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] On ZenHub, PR is connected to all issues it (partially) resolves
- [x] PR description links to connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] PR title references all connected issues
- [x] For each connected issue, there is at least one commit whose title references that issue

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title


### Author (partiality)

- [x] Added `p` tag to titles of partial commits
- [x] This PR is labeled `partial` <sub>or completely resolves all connected issues</sub>
- [x] This PR partially resolves each of the connected issues <sub>or does not have the `partial` label</sub>


### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or is not chained to another PR</sub>
- [x] The blocking PR is labeled `base` <sub>or this PR is not chained to another PR</sub>
- [x] This PR is labeled `chained` <sub>or is not chained to another PR</sub>


### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or the changes introduced by this PR will not require reindexing of any deployment</sub>
- [x] This PR is labeled `reindex:dev` <sub>or the changes introduced by it will not require reindexing of `dev`</sub>
- [x] This PR is labeled `reindex:anvildev` <sub>or the changes introduced by it will not require reindexing of `anvildev`</sub>
- [x] This PR is labeled `reindex:anvilprod` <sub>or the changes introduced by it will not require reindexing of `anvilprod`</sub>
- [x] This PR is labeled `reindex:prod` <sub>or the changes introduced by it will not require reindexing of `prod`</sub>
- [x] This PR is labeled `reindex:partial` and its description documents the specific reindexing procedure for `dev`, `anvildev`, `anvilprod` and `prod` <sub>or requires a full reindex or carries none of the labels `reindex:dev`, `reindex:anvildev`, `reindex:anvilprod` and `reindex:prod`</sub>
- [x] This PR and its connected issues are labeled `API` <sub>or this PR does not modify a REST API</sub>
- [x] Added `a` (`A`) tag to commit title for backwards (in)compatible changes <sub>or this PR does not modify a REST API</sub>
- [x] Updated REST API version number in `app.py` <sub>or this PR does not modify a REST API</sub>


### Author (upgrading deployments)

- [x] Ran `make image_manifests.json` and committed the resulting changes <sub>or this PR does not modify `azul_docker_images`, or any other variables referenced in the definition of that variable</sub>
- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading deployments</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading deployments</sub>
- [x] This PR is labeled `upgrade` <sub>or does not require upgrading deployments</sub>
- [x] This PR is labeled `deploy:shared` <sub>or does not modify `image_manifests.json`, and does not require deploying the `shared` component for any other reason</sub>
- [x] This PR is labeled `deploy:gitlab` <sub>or does not require deploying the `gitlab` component</sub>
- [x] This PR is labeled `deploy:runner` <sub>or does not require deploying the `runner` image</sub>


### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the none of the stable branches (`anvilprod` and `prod`) have temporary hotfixes for any of the issues connected to this PR</sub>


### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not modify `requirements*.txt`, `common.mk`, `Makefile` and `Dockerfile`</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not modify `requirements*.txt`</sub>
- [x] This PR is labeled `reqs` <sub>or does not modify `requirements*.txt`</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not modify functionality that could affect the IT outcome</sub>


### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] PR is awaiting requested review from system administrator
- [x] PR is assigned to only the system administrator


### System administrator (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] A comment to this PR details the completed security design review
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] PR is assigned to only the operator


### Operator (before pushing merge the commit)

- [x] Checked `reindex:…` labels and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Ran `_select dev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select dev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Ran `_select anvildev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Checked the items in the next section <sub>or this PR is labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the system administrator <sub>or this PR is not labeled `deploy:gitlab`</sub>


### System administrator

- [x] Background migrations for `dev.gitlab` are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Background migrations for `anvildev.gitlab` are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the operator


### Operator (before pushing merge the commit)

- [x] Ran `_select dev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Ran `_select anvildev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `dev`</sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `anvildev`</sub>
- [x] Started reindex in `sandbox` <sub>or this PR is not labeled `reindex:dev`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR is not labeled `reindex:anvildev`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR is not labeled `reindex:dev`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR is not labeled `reindex:anvildev`</sub>
- [x] The title of the merge commit starts with the title of this PR
- [x] Added PR # reference to merge commit title
- [x] Collected commit title tags in merge commit title <sub>but only included `p` if the PR is also labeled `partial`</sub>
- [x] Moved connected issues to *Merged lower* column in ZenHub
- [x] Moved blocked issues to *Triage* <sub>or no issues are blocked on the connected issues</sub>
- [x] Pushed merge commit to GitHub


### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>


### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev`
- [x] Pushed merge commit to GitLab `anvildev`
- [x] Build passes on GitLab `dev`
- [x] Reviewed build logs for anomalies on GitLab `dev`
- [x] Build passes on GitLab `anvildev`
- [x] Reviewed build logs for anomalies on GitLab `anvildev`
- [x] Ran `_select dev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`


### Operator (reindex)

- [x] Deindexed all unreferenced catalogs in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Deindexed all unreferenced catalogs in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Deindexed specific sources in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Deindexed specific sources in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Indexed specific sources in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Indexed specific sources in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Checked for, triaged and possibly requeued messages in both fail queues in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Checked for, triaged and possibly requeued messages in both fail queues in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Emptied fail queues in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Emptied fail queues in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>


### Operator

- [x] Propagated the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `API`, `reindex:partial`, `reindex:anvilprod` and `reindex:prod` labels to the next promotion PRs <sub>or this PR carries none of these labels</sub>
- [x] Propagated any specific instructions related to the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `API`, `reindex:partial`, `reindex:anvilprod` and `reindex:prod` labels, from the description of this PR to that of the next promotion PRs <sub>or this PR carries none of these labels</sub>
- [x] PR is assigned to no one


## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

dsotirho-ucsc committed:
Fix: 4xx and 5xx responses from API Gateway cause Invicti findings (DataBiosphere/azul#154)

dsotirho-ucsc commented:
[6385_IT_2024-08-19.txt](https://github.com/user-attachments/files/16666466/6385_IT_2024-08-19.txt)

dsotirho-ucsc commented:
[6385_IT_2024-08-19.txt](https://github.com/user-attachments/files/16666758/6385_IT_2024-08-19.txt)

achave11-ucsc reviewed (commented):
Please rebase this branch on the latest `develop`.

dsotirho-ucsc commented:
[6385_IT_2024-08-20.txt](https://github.com/user-attachments/files/16684858/6385_IT_2024-08-20.txt)

achave11-ucsc reviewed (dismissed):
✅

hannes-ucsc commented on src/azul/terraform.py:
```diff
@@ -800,11 +800,32 @@ def tf_config(self, app_name):
         # Setting this property using AWS API Gateway extensions to the OpenAPI
         # specification works around this issue.
         #
+        # We also use AWS API extensions to set the API Gateway default
+        # responses since, unlike Terraform, it automatically triggers the
+        # deployment of the REST API when the Gateway responses are added and/or
+        # modified.
+        #
```
```suggestion
        # We ran into similar difficulties when using Terraform to configure default responses for the API, so we use these extensions for that purpose, too.
        #
```

hannes-ucsc commented on src/azul/terraform.py:
```diff
@@ -800,11 +800,32 @@ def tf_config(self, app_name):
+            f'DEFAULT_{response_type}': {
+                'responseParameters': {
+                    f'gatewayresponse.header.{k}': v
+                    for k, v in [
```
Using a dictionary literal saves two characters per entry and makes it obvious that the keys are unique.

hannes-ucsc commented on src/azul/terraform.py:
```diff
@@ -800,11 +800,32 @@ def tf_config(self, app_name):
+                'responseParameters': {
+                    f'gatewayresponse.header.{k}': v
+                    for k, v in [
+                        ('Content-Security-Policy', "'default-src \'self\''"),
```
That literal is the same as this one: `"'default-src 'self''"`.

hannes-ucsc commented on src/azul/terraform.py:
```diff
@@ -800,11 +800,32 @@ def tf_config(self, app_name):
+                    for k, v in [
+                        ('Content-Security-Policy', "'default-src \'self\''"),
+                        ('X-Content-Type-Options', "'nosniff'"),
+                        ('X-Frame-Options', "'DENY'"),
```
This looks odd. A comment should explain why every value is enclosed in single quotes. You may also want to use the patterns that we followed for CSP in the browser TF component and the app.

dsotirho-ucsc commented:
[6385_IT_2024-08-27.txt](https://github.com/user-attachments/files/16766030/6385_IT_2024-08-27.txt)

hannes-ucsc commented on src/azul/terraform.py:
```diff
@@ -793,18 +793,50 @@ def tf_config(self, app_name):
         del deployment['lifecycle']
         deployment['triggers'] = {'redeployment': deployment.pop('stage_description')}
 
+        # FIXME: Replace with single_quote() from strings.py
```
FIXMEs can't refer to PRs. This should be a drop commit and the PRs should be chained. Or you just wait with the review request until the blocking PR has been merged.

hannes-ucsc commented on src/azul/terraform.py:
```diff
@@ -793,18 +793,50 @@ def tf_config(self, app_name):
+            f'DEFAULT_{response_type}': {
+                'responseParameters': {
+                    f'gatewayresponse.header.{k}': v
+                    for k, v in {
+                        # Static value method response header parameters must be
+                        # enclosed within a pair of single quotes.
+                        #
+                        # https://docs.aws.amazon.com/apigateway/latest/developerguide/request-response-data-mappings.html#mapping-response-parameters
+                        # https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-swagger-extensions-gateway-responses.html
+                        #
+                        'Content-Security-Policy': "'default-src 'self''",  # internal quotes intentionally not escaped
+                        'X-Content-Type-Options': sq('nosniff'),
+                        'X-Frame-Options': sq('DENY'),
+                        'Referrer-Policy': sq('strict-origin-when-cross-origin'),
+                        'Strict-Transport-Security': sq('max-age=63072000; includeSubDomains; preload'),
+                        'X-XSS-Protection': sq('1; mode=block')
+                    }.items()
```
Extract the dictionary literal into a variable and apply the quoting function to each value in the comprehension. The quoting function should allow internal quotes, or you don't use a quoting function at all.

dsotirho-ucsc replied:
Since it makes sense in all other cases for the quoting function to disallow single quotes in the value the it is wrapping in single quotes, I opted to not use the quoting function here and instead explain with a brief comment.

dsotirho-ucsc commented:
[6385_IT_2024-08-29.txt](https://github.com/user-attachments/files/16802772/6385_IT_2024-08-29.txt)

hannes-ucsc commented on src/azul/terraform.py:
```diff
@@ -800,11 +800,42 @@ def tf_config(self, app_name):
+                    # https://docs.aws.amazon.com/apigateway/latest/developerguide/request-response-data-mappings.html#mapping-response-parameters
+                    # https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-swagger-extensions-gateway-responses.html
+                    #
+                    # Note that single_quote() is not used here since the value
+                    # can contain internal single quotes. Testing found that the
+                    # internal quotes should not be escaped. e.g. "'foo 'bar''"
```
```suggestion
                    # Note that azul.strings.single_quote() is not used here since API Gateway allows internal single quotes in the value, which that function would prohibit.
```

hannes-ucsc commented on test/integration_test.py:
```diff
@@ -2093,3 +2093,18 @@ def test_response_security_headers(self):
                         self.assertIsSubset(expected.items(), response.headers.items())
+
+    def test_default_4xx_response_headers(self):
+        headers = {
```
This is (or should be) identical to the above dictionary literal. Please extract the above variable to a class attribute and import and use that here.

Furthermore, there are a bunch of related literals in similar integration tests (look for global_headers) and in chalice.py. I think these can and should be backed by that same attribute. Please perform this refactoring in a separate commit.

dsotirho-ucsc committed:
Adjust Strict-Transport-Security header value

dsotirho-ucsc commented:
[6385_IT_2024-09-10.txt](https://github.com/user-attachments/files/16955880/6385_IT_2024-09-10.txt)

hannes-ucsc commented on src/azul/headers.py:
```diff
@@ -0,0 +1,11 @@
+class ResponseHeaders:
```
Nah. This isn't worth a module of its own. Raise in PL and be ready to reproduce the original import error.

hannes-ucsc commented on src/azul/headers.py:
```diff
@@ -0,0 +1,11 @@
+class ResponseHeaders:
+    seconds = 60 * 60 * 24 * 365 * 2
```
```suggestion
    hsts_max_age = 60 * 60 * 24 * 365 * 2
```

dsotirho-ucsc committed:
Add additional security headers

dsotirho-ucsc commented:
[6385_IT_2024-09-12.txt](https://github.com/user-attachments/files/16987216/6385_IT_2024-09-12.txt)

Adding fixup on top of the previous refactor got a bit messy, so I reverted the commit and added a new refactor commit instead.

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -189,15 +189,30 @@ def _api_gateway_context_middleware(self, event, get_response):
 
+    hsts_max_age = 60 * 60 * 24 * 365 * 2
+
+    response_headers = {
```
These are all security-related headers so the name of the variable should reflect that.

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -189,15 +189,30 @@ def _api_gateway_context_middleware(self, event, get_response):
+    hsts_max_age = 60 * 60 * 24 * 365 * 2
+
+    response_headers = {
+        'Content-Security-Policy': "default-src 'self'",
+        'X-Content-Type-Options': 'nosniff',
+        'X-Frame-Options': 'DENY',
+        'Referrer-Policy': 'strict-origin-when-cross-origin',
+        'Strict-Transport-Security': f'max-age={hsts_max_age};'
+                                     ' includeSubDomains;'
+                                     ' preload',
+        'X-XSS-Protection': '1; mode=block'
```
This should use the string functions for joining words and so on that you added in some other PR, otherwise we end up with inconsistencies again.

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -189,15 +189,30 @@ def _api_gateway_context_middleware(self, event, get_response):
-        response.headers['Strict-Transport-Security'] = f'max-age={seconds}; includeSubDomains'
-        response.headers['X-Content-Type-Options'] = 'nosniff'
-        response.headers['X-Frame-Options'] = 'DENY'
+        for header in [
+            'Strict-Transport-Security',
+            'X-Content-Type-Options',
+            'X-Frame-Options',
+        ]:
```
Now you're duplicating this literal which partially defeats the purpose again.

Furthermore, it is unclear as to why you only specify some of the headers here. It seems like you should segregate the central definition according to purpose so that we don't end up leaking those details into the usage sites. The segregation will make the intent clear.

dsotirho-ucsc commented:
[6385_IT_2024-09-16.txt](https://github.com/user-attachments/files/17016423/6385_IT_2024-09-16.txt)

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -189,15 +189,33 @@ def _api_gateway_context_middleware(self, event, get_response):
+    hsts_max_age = 60 * 60 * 24 * 365 * 2
+
+    security_headers = {
+        # FIXME: Use join_words()
+        #        https://github.com/DataBiosphere/azul-private/issues/6
```
Just move the necessary commits to this PR.

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -189,15 +189,33 @@ def _api_gateway_context_middleware(self, event, get_response):
+        'X-Frame-Options': 'DENY',
+    }
+
+    default_response_headers = {
```
A comment should explain why they are segregated.

hannes-ucsc replied:
The name of the attribute contradicts the comment. Given the name of the variable, one would expect a "default" to be included in other definitions. You include the other definition in the "default".

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -189,15 +189,33 @@ def _api_gateway_context_middleware(self, event, get_response):
+
+    default_response_headers = {
+        **security_headers,
+        # FIXME: Use join_words() and single_quote()
```
Just move the necessary commits to this PR.

dsotirho-ucsc committed:
Refactor string manipulation functions into strings.py

dsotirho-ucsc commented:
[6385_IT_2024-09-17.txt](https://github.com/user-attachments/files/17032621/6385_IT_2024-09-17.txt)

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -189,15 +193,35 @@ def _api_gateway_context_middleware(self, event, get_response):
+        'X-Frame-Options': 'DENY',
+    }
+
+    # These headers are added to 4XX and 5XX responses from the API Gateway, for
+    # example if an invalid URL was requested (404), or if a request took too
+    # long to complete (504).
```
```suggestion
    # These headers are added to 4XX and 5XX responses from API Gateway, for
    # example if an invalid URL was requested (404), or if a request took too
    # long to complete (504).
```

It's still not clear from these comments as to why these definitions are separate. You explain the "what" and the "how", but not they "why": Why can't `Content-Security-Policy`, `Referrer-Policy` and `X-XSS-Protection` be part of all Lambda-generated responses?

dsotirho-ucsc replied:
I see no reason why `Referrer-Policy` and `X-XSS-Protection` can't be included in all responses. `Content-Security-Policy` however (in it's current form) is only suitable for an error response, so will remain separate. (https://github.com/DataBiosphere/azul-private/issues/6 will follow this PR to add the CSP header to all responses.)
In the latest fixup commit I've moved the two headers mentioned above, and rewritten the comments.

dsotirho-ucsc commented:
[6385_IT_2024-09-18.txt](https://github.com/user-attachments/files/17062623/6385_IT_2024-09-18.txt)

dsotirho-ucsc commented:
[6385_IT_2024-09-19.txt](https://github.com/user-attachments/files/17067636/6385_IT_2024-09-19.txt)

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -189,15 +193,37 @@ def _api_gateway_context_middleware(self, event, get_response):
+    #
+    api_gateway_4xx_5xx_headers = {
+        **security_headers,
+        'Content-Security-Policy': jw('default-src', sq('self')),
```
I assume this CSP is suitable for all endpoints and pages, except the Swagger login stuff. You should code it that way. Move the CSP header to `security_headers` and add an exception to drop the header where it is problematic. Prefix the exception with a FIXME referring to the issue that adds the nonce CSP.

hannes-ucsc commented:
I don't need fixups next time.

dsotirho-ucsc committed:
Add FIXME (DataBiosphere/azul-private#6)

dsotirho-ucsc commented:
[6385_IT_2024-09-25.txt](https://github.com/user-attachments/files/17136061/6385_IT_2024-09-25.txt)

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -189,15 +193,33 @@ def _api_gateway_context_middleware(self, event, get_response):
 
+    hsts_max_age = 60 * 60 * 24 * 365 * 2
+
+    # Headers added to every response from an AWS Lambda function, as well as
+    # 4XX and 5XX responses from the API. Use of these headers prevent known
+    # security vulnerabilities.
+    #
```
```suggestion
    # Headers added to every response from the app, as well as
    # canned 4XX and 5XX responses from API Gateway. Use of these headers addresses known
    # security vulnerabilities.
    #
```

hannes-ucsc reviewed (changes requested):
Correct me if I am wrong but commit 3 breaks Swagger UI and commit 4 fixes that. If that's the case, they should be a split commit.

I noticed that I've gotten two review requests on this normal priority PR, while I got none on the high priority #6110.

dsotirho-ucsc commented:
> Correct me if I am wrong but commit 3 breaks Swagger UI and commit 4 fixes that.

I've slightly re-arranged the commits, and instead of making a split commit simply merged the fix of the break in with the introduction of the break.

dsotirho-ucsc commented:
[6385_IT_2024-09-30.txt](https://github.com/user-attachments/files/17197509/6385_IT_2024-09-30.txt)

nadove-ucsc commented on src/azul/chalice.py:
```diff
@@ -195,8 +195,8 @@ def _api_gateway_context_middleware(self, event, get_response):
-    # Headers added to every response from an AWS Lambda function, as well as
-    # 4XX and 5XX responses from the API. Use of these headers prevent known
+    # Headers added to every response from the app, as well as canned 4XX and
+    # 5XX responses from API Gateway. Use of these headers addresses known
```
The non-semantic changes should be moved to a `fixup` that immediately follows the 3rd commit (where this paragraph was first added).

nadove-ucsc reviewed (changes requested):
The 3rd commit should be split into two commits, as implied by its title: first, a refactoring commit, then a 2nd commit to add the new headers. I understand that the current history was designed to avoid temporarily breaking the Swagger UI, but a refactoring is semantically neutral by definition, so my proposal shouldn't break anything.

dsotirho-ucsc committed:
Refactor security headers in AzulChaliceApp

dsotirho-ucsc commented:
[6385_IT_2024-10-01.txt](https://github.com/user-attachments/files/17218635/6385_IT_2024-10-01.txt)

nadove-ucsc commented:
Looks good to me but I'd like @hannes-ucsc to give final approval.

nadove-ucsc commented:
### Security design review

- [ ] Security design review completed; this PR does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with Terra 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below
